server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -600,6 +603,13 @@ public void onFailure(Exception e) {
         if (newRouting.equals(currentRouting) == false) {
             indexEventListener.shardRoutingChanged(this, currentRouting, newRouting);
         }
+
+        if (indexSettings.isSoftDeleteEnabled() && useRetentionLeasesInPeerRecovery == false) {
+            if (getPeerRecoveryRetentionLeases().size() >= routingTable.size()) {
```

Comment:
Should we check that we have peer recovery retention leases for all shards in the routing table or is that not necessary?

Reply:
Yeah, we discussed this a week ago and preferred the PRRL size check approach. However, we might turn off translog too soon in case of shard relocations. If we turn off the translog retention policy only after all shards have peer recovery retention leases, then the upgrade path will be smoother. The downside of the new approach is that the translog retention policy can resurrect in a full cluster restart for a short period. We also agreed that it would not be a big deal. I adjusted this in https://github.com/elastic/elasticsearch/pull/49448/commits/d74759078e2e93777d971b3ba6f507430aaba100.

Reply:
I think that the adapted logic in d747590 does not properly handle relocating shards. `StreamSupport.stream(routingTable.spliterator(), false)` does not return relocation targets as far as I'm aware.

Reply:
I think we don't need to handle relocating shards. Anyway, it's safer to include them. I pushed https://github.com/elastic/elasticsearch/pull/49448/commits/55819238a5c3791e970f947cd097809850f09ce7.